File: pingProcessorGolang/controller/sensor_controller.go
package controller

import (
	"encoding/json"
	"log"
	"net/http"
	"pingProcessorGolang/response"
	"pingProcessorGolang/service"
	"strconv"
)

type SensorController struct {
	sensorService service.SensorService
}

func NewSensorController(sensorService service.SensorService) *SensorController {
	return &SensorController{sensorService: sensorService}
}

func (s *SensorController) SendPing(w http.ResponseWriter, r *http.Request) {
	// Get the sensorID from the query parameters
	sensorIDStr := r.URL.Query().Get("sensorID")
	sensorID, err := strconv.Atoi(sensorIDStr)
	if err != nil {
		log.Printf("Invalid sensor ID: %v", err)
		response := response.ApiResponse{
			Success: false,
			Message: "Invalid sensor ID",
			Data:    nil,
		}
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(response)
		return
	}

	// Process the ping
	s.sensorService.ProcessPing(sensorID)
	log.Printf("Ping successfully processed for sensor ID: %d", sensorID)

	// Return success response
	response := response.ApiResponse{
		Success: true,
		Message: "Ping processed successfully",
		Data:    sensorID,
	}
	w.WriteHeader(http.StatusOK)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

func (s *SensorController) GetTopFrequencySensor(w http.ResponseWriter, r *http.Request) {
	// Get the 'k' parameter from the query parameters
	kStr := r.URL.Query().Get("k")
	k, err := strconv.Atoi(kStr)
	if err != nil || k <= 0 {
		log.Printf("Invalid range k: %v", err)
		response := response.ApiResponse{
			Success: false,
			Message: "Parameter 'k' must be greater than 0",
			Data:    nil,
		}
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(response)
		return
	}

	// Retrieve the sensor with the highest frequency
	topSensorID := s.sensorService.GetTopKMaxFreqSensorID(k)
	if topSensorID == -1 {
		log.Printf("No sensor data available for the given range: %d", k)
		response := response.ApiResponse{
			Success: false,
			Message: "No sensor data found for the given range",
			Data:    nil,
		}
		w.WriteHeader(http.StatusNotFound)
		json.NewEncoder(w).Encode(response)
		return
	}

	log.Printf("Top frequency sensor ID found: %d", topSensorID)

	// Return success response
	response := response.ApiResponse{
		Success: true,
		Message: "Top frequency sensor found",
		Data:    topSensorID,
	}
	w.WriteHeader(http.StatusOK)
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)
}

File: pingProcessorJava/src/main/java/com/project/sensor/pingProcessor/controller/SensorController.java
package com.project.sensor.pingProcessor.controller;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.project.sensor.pingProcessor.service.SensorService;
import com.project.sensor.pingProcessor.response.ApiResponse;

@RestController
@RequestMapping("/api/sensors")
public class SensorController {

    private static final Logger logger = LoggerFactory.getLogger(SensorController.class);
    private final SensorService sensorService;

    @Autowired
    public SensorController(SensorService sensorService) {
        this.sensorService = sensorService;
    }

    @PostMapping("/ping")
    public ResponseEntity<ApiResponse> sendPing(@RequestParam int sensorID) {
        logger.info("Received ping request for sensor ID: {}", sensorID);
        try {
            sensorService.processPing(sensorID);
            logger.info("Ping successfully processed for sensor ID: {}", sensorID);
            ApiResponse response = new ApiResponse(true, "Ping processed successfully", sensorID);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            logger.error("Error processing ping for sensor ID: {}", sensorID, e);
            ApiResponse response = new ApiResponse(false, "Failed to process ping for sensor ID", null);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }

    @GetMapping("/top-frequency-sensor")
    public ResponseEntity<ApiResponse> getTopFrequencySensor(@RequestParam int k) {
        logger.info("Received request to get top-frequency sensor with range: {}", k);
        try {
            if (k <= 0) {
                logger.warn("Invalid range k: {}. Must be greater than 0.", k);
                ApiResponse response = new ApiResponse(false, "Parameter 'k' must be greater than 0", null);
                return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(response);
            }

            int topSensorID = sensorService.getTopKMaxFreqSensorID(k);
            if (topSensorID == -1) {
                logger.warn("No sensor data available for the given range: {}", k);
                ApiResponse response = new ApiResponse(false, "No sensor data found for the given range", null);
                return ResponseEntity.status(HttpStatus.NOT_FOUND).body(response);
            }

            logger.info("Top frequency sensor ID found: {}", topSensorID);
            ApiResponse response = new ApiResponse(true, "Top frequency sensor found", topSensorID);
            return ResponseEntity.ok(response);
        } catch (Exception e) {
            logger.error("Error retrieving top-frequency sensor for range: {}", k, e);
            ApiResponse response = new ApiResponse(false, "Failed to retrieve top-frequency sensor", null);
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(response);
        }
    }
}
